go/libraries/doltcore/sqle/database_provider.go

```diff
@@ -698,6 +706,45 @@ func (p DoltDatabaseProvider) DropDatabase(ctx *sql.Context, name string) error
 	}
 
 	delete(p.databases, dbKey)
+
+	return p.invalidateDbStateInAllSessions(ctx, name)
+}
+
+// invalidateDbStateInAllSessions removes the db state for this database from every session. This is necessary when a
+// database is dropped, so that other sessions don't use stale db state.
+func (p DoltDatabaseProvider) invalidateDbStateInAllSessions(ctx *sql.Context, name string) error {
+	// TODO: this might be a better fit as a function on SessionManager
+	// TODO: Instead of deleting the db state, could we just mark it as invalid and continue using that same instance
+	//       in memory, even when we create a new db with the same name? reusing for creation of a new db might be tricky.
+	runningServer := sqlserver.GetRunningServer()
```

Comment:
This global is super spooky! Can we thread the server through when we construct the database provider for the engine? Definitely not related to this PR per se, but seeing this global makes me want to immediately go try to remove it...

Reply:
Well... this turned out to be a little messier than I was expecting... It looks like we're always using `DoltDatabaseProvider` as a struct for some reason, and not as a pointer to a struct. The `DoltDatabaseProvider` struct gets created when we create the engine and then passed (as a copy) into GMS. After that, the `sqlEngine` is instantiated and we can start the sql-server at which point we get the server reference that we want to set in the provider, but we can't change the provider registered in GMS, because a copy was passed in, and we don't have a reference to that memory. 

I'm happy to change `DoltDatabaseProvider` to be used as a pointer (seems like a good change to make for several reasons), but it feels like that might be cleaner in a separate PR since it'll be more than just a couple real quick changes like I thought this was going to be.

Reply:
Thanks for the explanation and investigating this. I agree this doesn't need to be addressed here :).